"use strict";(self.webpackChunkdocs=self.webpackChunkdocs||[]).push([[4114],{3905:(e,t,n)=>{n.d(t,{Zo:()=>c,kt:()=>f});var r=n(7294);function i(e,t,n){return t in e?Object.defineProperty(e,t,{value:n,enumerable:!0,configurable:!0,writable:!0}):e[t]=n,e}function o(e,t){var n=Object.keys(e);if(Object.getOwnPropertySymbols){var r=Object.getOwnPropertySymbols(e);t&&(r=r.filter((function(t){return Object.getOwnPropertyDescriptor(e,t).enumerable}))),n.push.apply(n,r)}return n}function a(e){for(var t=1;t<arguments.length;t++){var n=null!=arguments[t]?arguments[t]:{};t%2?o(Object(n),!0).forEach((function(t){i(e,t,n[t])})):Object.getOwnPropertyDescriptors?Object.defineProperties(e,Object.getOwnPropertyDescriptors(n)):o(Object(n)).forEach((function(t){Object.defineProperty(e,t,Object.getOwnPropertyDescriptor(n,t))}))}return e}function s(e,t){if(null==e)return{};var n,r,i=function(e,t){if(null==e)return{};var n,r,i={},o=Object.keys(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||(i[n]=e[n]);return i}(e,t);if(Object.getOwnPropertySymbols){var o=Object.getOwnPropertySymbols(e);for(r=0;r<o.length;r++)n=o[r],t.indexOf(n)>=0||Object.prototype.propertyIsEnumerable.call(e,n)&&(i[n]=e[n])}return i}var l=r.createContext({}),p=function(e){var t=r.useContext(l),n=t;return e&&(n="function"==typeof e?e(t):a(a({},t),e)),n},c=function(e){var t=p(e.components);return r.createElement(l.Provider,{value:t},e.children)},u="mdxType",m={inlineCode:"code",wrapper:function(e){var t=e.children;return r.createElement(r.Fragment,{},t)}},d=r.forwardRef((function(e,t){var n=e.components,i=e.mdxType,o=e.originalType,l=e.parentName,c=s(e,["components","mdxType","originalType","parentName"]),u=p(n),d=i,f=u["".concat(l,".").concat(d)]||u[d]||m[d]||o;return n?r.createElement(f,a(a({ref:t},c),{},{components:n})):r.createElement(f,a({ref:t},c))}));function f(e,t){var n=arguments,i=t&&t.mdxType;if("string"==typeof e||i){var o=n.length,a=new Array(o);a[0]=d;var s={};for(var l in t)hasOwnProperty.call(t,l)&&(s[l]=t[l]);s.originalType=e,s[u]="string"==typeof e?e:i,a[1]=s;for(var p=2;p<o;p++)a[p]=n[p];return r.createElement.apply(null,a)}return r.createElement.apply(null,n)}d.displayName="MDXCreateElement"},1301:(e,t,n)=>{n.r(t),n.d(t,{assets:()=>l,contentTitle:()=>a,default:()=>m,frontMatter:()=>o,metadata:()=>s,toc:()=>p});var r=n(7462),i=(n(7294),n(3905));const o={sidebar_position:7},a="Data Sampling",s={unversionedId:"visualizations/data-sampling",id:"visualizations/data-sampling",title:"Data Sampling",description:"For the efficient rendering of visualizations, entire data is not shown in the whole genome scale for point mutations and indels.",source:"@site/docs/visualizations/data-sampling.md",sourceDirName:"visualizations",slug:"/visualizations/data-sampling",permalink:"/docs/visualizations/data-sampling",draft:!1,editUrl:"https://github.com/hms-dbmi/chromoscope/tree/master/docs/docs/visualizations/data-sampling.md",tags:[],version:"current",sidebarPosition:7,frontMatter:{sidebar_position:7},sidebar:"docsSidebar",previous:{title:"Export",permalink:"/docs/visualizations/export"},next:{title:"Loading Data",permalink:"/docs/category/loading-data"}},l={},p=[],c={toc:p},u="wrapper";function m(e){let{components:t,...n}=e;return(0,i.kt)(u,(0,r.Z)({},c,n,{components:t,mdxType:"MDXLayout"}),(0,i.kt)("h1",{id:"data-sampling"},"Data Sampling"),(0,i.kt)("p",null,"For the efficient rendering of visualizations, entire data is not shown in the whole genome scale for point mutations and indels."),(0,i.kt)("p",null,"Chromoscope selects ",(0,i.kt)("strong",{parentName:"p"},"500")," ",(0,i.kt)("a",{parentName:"p",href:"#vcf--tbi"},"point mutations")," in each visible tile. Given that showing dense point mutations are important the corresponding tracks, Chromoscope keeps point mutations with relatively small ",(0,i.kt)("inlineCode",{parentName:"p"},"DISTPREV")," (i.e., distance to the previous mutation) and filters out a proportion mutations with relatively high ",(0,i.kt)("inlineCode",{parentName:"p"},"DISTPREV"),"."),(0,i.kt)("p",null,"Point mutations are selected dynamically based on the zoom level. When zooming in, more mutations will be displayed, until eventually, all point mutations become visible when zooming in close enough."))}m.isMDXComponent=!0}}]);